parsec/backend/shamir.py

```diff
@@ -0,0 +1,162 @@
+            brief_certificates.append(brief_certificate)
+            share_certificates.append(share_certificate)
+
+        return ShamirRecoveryOthersListRepOk(brief_certificates, share_certificates)
+
+    @api("shamir_recovery_self_info")
+    @catch_protocol_errors
+    @api_typed_msg_adapter(ShamirRecoverySelfInfoReq, ShamirRecoverySelfInfoRep)
+    async def api_recovery_self_info(
+        self, client_ctx: AuthenticatedClientContext, req: ShamirRecoverySelfInfoReq
+    ) -> ShamirRecoverySelfInfoRep:
+        self_info = await self.recovery_self_info(
+            organization_id=client_ctx.organization_id,
+            author=client_ctx.device_id,
+        )
+
+        return ShamirRecoverySelfInfoRepOk(self_info)
+
+    @api("shamir_recovery_setup")
+    @catch_protocol_errors
+    @api_typed_msg_adapter(ShamirRecoverySetupReq, ShamirRecoverySetupRep)
+    async def api_shamir_recovery_setup(
+        self, client_ctx: AuthenticatedClientContext, req: ShamirRecoverySetupReq
+    ) -> ShamirRecoverySetupRep:
+        try:
+            await self.recovery_setup(
+                organization_id=client_ctx.organization_id,
+                author=client_ctx.device_id,
+                author_verify_key=client_ctx.verify_key,
+                setup=req.setup,
+            )
+        except DataError:
+            return ShamirRecoverySetupRepInvalidData()
+
+        return ShamirRecoverySetupRepOk()
+
+    @api("invite_shamir_recovery_reveal", client_types=[ClientType.INVITED])
+    @catch_protocol_errors
+    @api_typed_msg_adapter(InviteShamirRecoveryRevealReq, InviteShamirRecoveryRevealRep)
+    async def api_invite_shamir_recovery_reveal(
+        self, client_ctx: InvitedClientContext, req: InviteShamirRecoveryRevealReq
+    ) -> InviteShamirRecoveryRevealRep:
+        ciphered_data = await self.recovery_reveal(
+            client_ctx.organization_id, reveal_token=req.reveal_token
+        )
+
+        if ciphered_data is None:
+            return InviteShamirRecoveryRevealRepNotFound()
+
+        return InviteShamirRecoveryRevealRepOk(ciphered_data)
+
+    async def recovery_others_list(
+        self,
+        organization_id: OrganizationID,
+        author: DeviceID,
+    ) -> list[tuple[bytes, bytes]]:
+        raise NotImplementedError()
+
+    async def recovery_self_info(
+        self,
+        organization_id: OrganizationID,
+        author: DeviceID,
+    ) -> bytes | None:
+        raise NotImplementedError()
+
+    async def recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: DeviceID,
+        author_verify_key: VerifyKey,
+        setup: ShamirRecoverySetup | None,
+    ) -> None:
+        raise NotImplementedError()
+
+    async def recovery_reveal(
+        self,
+        organization_id: OrganizationID,
+        reveal_token: ShamirRevealToken,
+    ) -> bytes | None:
+        raise NotImplementedError()
+
+    # Helpers
+
+    def _verify_certificates(
```

Comment:
This is the very first method called by `recovery_setup` hence it could be move in the api method.

I don't have strong opinion on where to put the check, but currently the checks are done in the api method (see what is done for `user_create` for instance)

The good thing about having the check at api level is it simplify the tests (typically when passing a serialized data as `b'dummy'` instead of valid-but-complex-to-generate stuff)

On the other hand, this means the internal methods are more dangerous to use (typically extra care must be taken if they get used in the cli)

I guess the ideal solution would be to pass to the function objects that do the validation when they are constructed (so that the test can hack around the validation), but this is probably not trivial to implement.

So do you think in the long run we should got your way for other validations (i.e. move the validation part from the api method to the internal method) ?

Reply:
That, and `shamir_recovery_setup` should be splitted in two

Reply:
[aac9a3a](https://github.com/Scille/parsec-cloud/pull/4504/commits/aac9a3a002ed398fd09120b4755fbdc367837029)

Reply:
[b8672b8](https://github.com/Scille/parsec-cloud/pull/4504/commits/b8672b87a6f9d5f4042711ef3e8d014653056b5e)

Reply:
[aac9a3a](https://github.com/Scille/parsec-cloud/pull/4504/commits/aac9a3a002ed398fd09120b4755fbdc367837029)